apps/src/dance/DanceAiClient.ts

```diff
@@ -0,0 +1,101 @@
+import UntypedCachedPalettes from '@cdo/static/dance/ai/model/cached-spacy-palette-map.json';
+import UntypedCachedBackgrounds from '@cdo/static/dance/ai/model/cached-spacy-background-map.json';
+import UntypedCachedForegrounds from '@cdo/static/dance/ai/model/cached-spacy-foreground-map.json';
+
+// Custom types for vector mapping JSONs to play nice with Typescript
+type EmojiAssociation = {
+  association: number[];
+};
+
+type CachedMapping = {
+  emojiAssociations: {[key: string]: EmojiAssociation};
+  output: string[];
+};
+
+const CachedPalettes: CachedMapping = UntypedCachedPalettes;
+const CachedBackgrounds: CachedMapping = UntypedCachedBackgrounds;
+const CachedForegrounds: CachedMapping = UntypedCachedForegrounds;
+
+// Output: JSON containing {backgroundColor:string, backgroundEffect:string, foregroundEffect:string}
+// ToDo: Break down into helper functions for each step rather than repeat code for each field
+export default function cachedAi(input: string) {
```

Comment:
Recommendation: rename this to "chooseEffects"
Reasoning: A good rule of thumb is to name functions with verbs ([useful cheat sheet I like](https://dev.to/maikomiyazaki/beginner-s-cheat-sheet-commonly-used-verbs-for-naming-functions-methods-and-variables-509i)). That way, it's clearer to the reader what action the function is carrying out. It also prevents the function from growing into doing something it shouldn't in the future. Having a noun here, "cachedAi" makes me think this is a variable and therefore will have the same value regardless of how I use it. Migrating to an action will clarify that confusion. This function also probably doesn't need the word "ai" in it since the file is named "DanceAiClient" and so it's known that this function is an action taken by an AI client (although I'm holding that opinion very loosely). Finally, I recommend against "cached" since that seems to be an implementation detail. Ideally, we'd have a "DanceAiClient" that would be created with the knowledge of whether it should use its cached version or query OpenAI directly. Then, later on, we'd call DanceAiClient.chooseEffects(emojis). DanceAiClient would have the freedom to act as an api client and query OpenAI or shortcut that process using its knowledge of a cache.

Reply:
Definitely agree with all of this, the function was named cachedAi to distinguish the source from the prior function that queryed ChatGPT named "doAi". We can just switch the naming conventions of both to be more indicative of what's actually going on though

Reply:
I think I created this file before we created a utils file and moved the existing `doAi` there:

https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/dance/ai/utils.ts

Might make sense to move this function into utils as well?

Reply:
How about we make migrating this into utils as a next-step task to help unblock other folks?

Reply:
Tracking this as a next step task in this Trello ticket: https://trello.com/c/KvfZFqlX